### PR TITLE
less repeated code

### PR DESCRIPTION
**Pull request by pre1ude**
I love the eventemitter3, it was written clearly and teached me a lot, but I found some code is repeated in it, so I try to merged some code. After run benchmark, I found the once benchmark is slower than it was, em.. function call is expensive so you inline it, but merge it, code will be more clearly and maintable

### CODE REVIEW

**Comment by lpinca** [comment]
Yes, there is some duplicated code and this is done for performance.

Instead of using `Function.prototype.call()` you can try pass the `thisArg` as a normal argument to a function. For example:

```js
function _addListener(emitter, event, fn, context, once) {
  // replace all occurences of `this` with `emitter`.
}
```

and then call it like

```js
EventEmitter.prototype.on = function on(event, fn, context) {
  return _addListener(this, event, fn, context, false);
};
```

This should have better performance as it avoids the `call()` overhead.

**Comment by pre1ude** [comment]
Thank you for your advice, and using your advice, performance is better but still half of the master version
```
EventEmitter3(my-fork) x 7,066,182 ops/sec ±3.94% (88 runs sampled)
EventEmitter3(master) x 13,128,453 ops/sec ±3.93% (89 runs sampled)
```

**Comment by pre1ude** [comment]
if there is any tools could help auto inline function before we minify it?

**Comment by lpinca** [comment]
How is performance if you don't minify it?
Can you post the results of `npm run benchmark`?

File: index.js [hunk 1]
```diff
@@ -165,119 +166,100 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
  * @returns {EventEmitter} `this`.
  * @api public
  */
 EventEmitter.prototype.removeListener = function removeListener(event, fn, context, once) {
-  var evt = prefix ? prefix + event : event;
+  var evt;
 
-  if (!this._events[evt]) return this;
-  if (!fn) {
+  function clearEvent() {
```
Comment (lpinca) [inline index.js]: Try to move this function outside the `removeListener` method, and call it without using `Function.prototype.call()` as you did above.

**Comment by pre1ude** [comment]
[performance](https://gist.github.com/pre1ude/27df0dbb2e49a3c62fb57635ad5ba2f7) before I made this commit
[performance](https://gist.github.com/pre1ude/922c5ff3fb37cbbf2735765effd64961) is better at least for `once` benchmark

**Comment by pre1ude** [comment]
I mean the tools helps inline the repeat invoked function and then minify it, maybe the performance will better

File: index.js [hunk 2]
```diff
@@ -166,118 +167,99 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
-
-  if (!this._events[evt]) this._events[evt] = listener, this._eventsCount++;
-  else if (!this._events[evt].fn) this._events[evt].push(listener);
-  else this._events[evt] = [this._events[evt], listener];
-
-  return this;
+  return _addListener(this, event, fn, context, true);
 };
 
+function clearEvent(emitter, evt) {
```
Comment (lpinca) [inline index.js]: JSDoc please.

File: index.js [hunk 3]
```diff
@@ -166,118 +167,99 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
  */
 EventEmitter.prototype.removeListener = function removeListener(event, fn, context, once) {
-  var evt = prefix ? prefix + event : event;
+  var evt;
 
-  if (!this._events[evt]) return this;
-  if (!fn) {
-    if (--this._eventsCount === 0) this._events = new Events();
-    else delete this._events[evt];
+  if(!event) {
```
Comment (lpinca) [inline index.js]: Please add a space after `if`.

File: index.js [hunk 4]
```diff
@@ -166,118 +167,99 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
-  if (listeners.fn) {
-    if (
-         listeners.fn === fn
-      && (!once || listeners.once)
-      && (!context || listeners.context === context)
-    ) {
-      if (--this._eventsCount === 0) this._events = new Events();
-      else delete this._events[evt];
-    }
   } else {
```
Comment (lpinca) [inline index.js]: This `else` branch is not needed as there is `return this;` in the corresponding `if` body. Please remove it.

File: index.js [hunk 5]
```diff
@@ -166,118 +167,99 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
-    for (var i = 0, events = [], length = listeners.length; i < length; i++) {
+    evt = prefix ? prefix + event : event;
+    if (!this._events[evt]) return this;
+    if(!fn) {
```
Comment (lpinca) [inline index.js]: Missing space after `if`.

File: index.js [hunk 6]
```diff
@@ -166,118 +167,99 @@ EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
...
-  else this._events[evt] = [this._events[evt], listener];
+  if (!emitter._events[evt]) emitter._events[evt] = listener, emitter._eventsCount++;
+  else if (!emitter._events[evt].fn) emitter._events[evt].push(listener);
+  else emitter._events[evt] = [emitter._events[evt], listener];
 
-  return this;
+  return emitter;
+}
+
+EventEmitter.prototype.on = function on(event, fn, context) {
```
Comment (pre1ude) [inline index.js]: but `_addListener` got the JSDoc already

File: index.js [hunk 7]
```diff
@@ -189,55 +204,65 @@ EventEmitter.prototype.on = function on(event, fn, context) {
...
  * @param {Boolean} once Only remove one-time listeners.
- * @returns {EventEmitter} `this`.
+ * @returns {EventEmitter} Reference to the EventEmitter instance.
  * @api public
  */
 EventEmitter.prototype.removeListener = function removeListener(event, fn, context, once) {
-  var evt = prefix ? prefix + event : event;
+  var evt;
 
+  if (!event) {
```
Comment (lpinca) [inline index.js]: This is a breaking change.

```js
ee.removeListener();
```

does not remove all events when no event is specified.

**Comment by lpinca** [comment]
There are some nits in JSDoc comments which can be fixed later.
Is there any performance regression compared to master?

**Comment by pre1ude** [comment]
I made a comparison result of the performance, [perf-regression.md](https://gist.github.com/pre1ude/5b41ef5f62bbdf27b0e7fef58bea434f)
as we can see, the perf is almost the same as the master

File: index.js [hunk 8]
```diff
@@ -260,24 +279,16 @@ EventEmitter.prototype.removeListener = function removeListener(event, fn, conte
...
-    }
-  } else {
+  if (!event) {
     this._events = new Events();
     this._eventsCount = 0;
+    return this;
   }
-
-  return this;
+  return this.removeListener(event);
```
Comment (lpinca) [inline index.js]: On second thought I think this isn't a good idea as we are now incurring in all the overhead need to filter the listeners (the for loop).

I bet `ee.removeAllListeners('foo')` is way slower now.

Comment (pre1ude) [inline index.js]: yeap, more slower..., I shouldn't modify the `removeAllListeners`.

**Comment by lpinca** [comment]
Thanks.

**Comment by pre1ude** [comment]
My pleasure :)

**Pull request merged**